### PR TITLE
Add new functionality to MonitorTxError

### PR DESCRIPTION
**Pull request by DavidZagury**
**What I did**
 Added class in orchagent named MonitorTxError - This class will monitor ports TX error counters, and will change the port status if the TX error is above a threshold
 Added it to the orchs list in orchdaemon class
**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

File: orchagent/MonitorTxError.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+}
+
+
+void MonitorTxError::doTask(Consumer& consumer)
+{
+	SWSS_LOG_ENTER();
+	string tableName = consumer.getTableName();
+
+	SWSS_LOG_NOTICE("DEDDY MonitorTxError doTask consumer tableName %s\n", tableName.c_str());
```
Comment (DavidZagury) [inline orchagent/MonitorTxError.cpp:64]: Remove DEDDY from logs

Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Move to debug and remove Deddy

File: orchagent/MonitorTxError.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+				SWSS_LOG_WARN("Cannot change pooling_period to the value entered.");
+			}
+		}
+		else
+		{
+			SWSS_LOG_WARN("Unknown field value");
+		}
+	}
+
+}
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Add new line at eof

File: orchagent/MonitorTxError.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+		if (p.m_type != Port::PHY)
+		{
+			continue;
+		}
+
+		string oidStr;
+		if (!m_countersPortNameMap->hget("", name, oidStr))
+		{
+			SWSS_LOG_ERROR("DEDDY error getting port name from counters");
+			return;
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Change return to continue

File: orchagent/MonitorTxError.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+		}
+		updatePortState(oidStr, p);
+
+	}
+}
+
+void MonitorTxError::updatePortState(string oidStr, Port port)
+{
+	PortState portState = OK;
+
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Remove enter

File: orchagent/MonitorTxError.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+		{
+			SWSS_LOG_ERROR("DEDDY error getting port name from counters");
+			return;
+		}
+		updatePortState(oidStr, p);
+
+	}
+}
+
+void MonitorTxError::updatePortState(string oidStr, Port port)
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Use reference for port

File: orchagent/MonitorTxError.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+	string txErrStrValue;
+	uint64_t txErrValue;
+	uint64_t txErrDiff;
+	if (!m_countersTable->hget(oidStr, counterName, txErrStrValue))
+	{
+		portState = UNKNOWN;
+		SWSS_LOG_ERROR("Cannot take information from table for port %s", oidStr.c_str());
+		return;
+	}
+	txErrValue = stoul(txErrStrValue);
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Use try catch sice stoul is not safe

File: orchagent/MonitorTxError.cpp [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+#include "port.h"
+#include "select.h"
+#include "portsorch.h"
+#include "sai_serialize.h"
+#include <array>
+
+using namespace std;
+using namespace swss;
+
+
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Remove redundant enters

File: orchagent/MonitorTxError.cpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+	initTimer(TX_ERROR_CHECK_DEFAULT_TIMER);
+
+	vector <FieldValueTuple> fv;
+	fv.emplace_back(VALUE, to_string(m_poolingPeriod));
+	m_txErrConfigTable.set(POOLING_PERIOD_KEY, fv);
+	fv.clear();
+	fv.emplace_back(VALUE, to_string(m_threshold));
+	m_txErrConfigTable.set(THRESHOLD_KEY, fv);
+
+	SWSS_LOG_NOTICE("DEDDY MonitorTxError initialized\n");
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: move to debug

File: orchagent/MonitorTxError.cpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+	m_poolingPeriod = setPoolingPeriod;
+	auto interv = timespec{ .tv_sec = m_poolingPeriod, .tv_nsec = 0 };
+	timer = new SelectableTimer(interv);
+	auto executor = new ExecutableTimer(timer, this, "TX_ERROR_POOLING_PERIOD");
+	Orch::addExecutor(executor);
+	SWSS_LOG_NOTICE("Initializing timer with %d seconds ...", m_poolingPeriod);
+	timer->start();
+}
+
+
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Remove enter

File: orchagent/MonitorTxError.cpp [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+			}
+		}
+		else
+		{
+			SWSS_LOG_WARN("Unknown field value");
+		}
+	}
+}
+
+
```
Comment (liorghub) [inline orchagent/MonitorTxError.cpp]: Remove enter

File: orchagent/MonitorTxError.h [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+#include "selectabletimer.h"
+#include "select.h"
+#include "timer.h"
+
+using namespace std;
+using namespace swss;
+extern "C"
+{
+#include "sai.h"
+}
```
Comment (liorghub) [inline orchagent/MonitorTxError.h]: why do we need this include?

File: orchagent/MonitorTxError.h [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+{
+public:
+	MonitorTxError(TableConnector configDBTConnector, TableConnector stateDBTConnector);
+
+	virtual void doTask(swss::SelectableTimer& timer);
+
+	virtual void doTask(Consumer& consumer);
+
+	virtual ~MonitorTxError(void);
+
```
Comment (liorghub) [inline orchagent/MonitorTxError.h:35]: Remove enters

File: orchagent/MonitorTxError.h [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+	shared_ptr <swss::Table> m_countersTable = nullptr;
+	shared_ptr <Table> m_countersPortNameMap = nullptr;
+
+	// state table
+	Table m_stateTxErrTable;
+	Table m_txErrConfigTable;
+
+	map <string, string> m_portsStringsMap;
+
+
```
Comment (liorghub) [inline orchagent/MonitorTxError.h]: Remove enters

File: orchagent/MonitorTxError.h [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+	// counters table - taken from redis
+	shared_ptr <swss::DBConnector> m_countersDB = nullptr;
+	shared_ptr <swss::Table> m_countersTable = nullptr;
+	shared_ptr <Table> m_countersPortNameMap = nullptr;
+
+	// state table
+	Table m_stateTxErrTable;
+	Table m_txErrConfigTable;
+
+	map <string, string> m_portsStringsMap;
```
Comment (liorghub) [inline orchagent/MonitorTxError.h]: m_portsStringsMap, do we need it?

File: orchagent/MonitorTxError.h [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+
+	// counetrs of tx errors
+	map <string, uint64_t> m_portsStateTable;
+	map <string, uint64_t> m_lastTxCounters;
+	// map<string, uint64_t> m_currTxCounters;
```
Comment (liorghub) [inline orchagent/MonitorTxError.h]: Remove comment

File: orchagent/MonitorTxError.h [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+	map <string, PortState> m_portsState;
+
+	uint64_t m_threshold = TX_ERROR_DEFAULT_THRESHOLD;
```
Comment (liorghub) [inline orchagent/MonitorTxError.h]: Add _sec to variable name

File: orchagent/MonitorTxError.h [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+	int m_poolingPeriod = TX_ERROR_CHECK_DEFAULT_TIMER;
```
Comment (liorghub) [inline orchagent/MonitorTxError.h]: Add _packets to var name

File: orchagent/MonitorTxError.h [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+	void updateAllPortsState();
+
+	void updatePortState(string portAlias, Port port);
+
+	void setThreshold(const vector <FieldValueTuple>& data);
+
+	void setPoolingPeriod(const vector <FieldValueTuple>& data);
+
+	void setDefaultConfigParam();
+//    static void updateConfig(SyncMap op);
```
Comment (liorghub) [inline orchagent/MonitorTxError.h]: Remove all enters and updateConfig

File: orchagent/MonitorTxError.h [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+};
+
+
+#endif //SONIC_SWSS_MONITORTXERROR_H
```
Comment (liorghub) [inline orchagent/MonitorTxError.h:62]: Remove all enters and add one enter at eof.

File: orchagent/orchdaemon.cpp [hunk 20]
```diff
@@ -9,6 +9,7 @@
 #define SAI_SWITCH_ATTR_CUSTOM_RANGE_BASE SAI_SWITCH_ATTR_CUSTOM_RANGE_START
 #include "sairedis.h"
 #include "chassisorch.h"
+#include "MonitorTxError.h"
```
Comment (liorghub) [inline orchagent/orchdaemon.cpp]: Move to orchdaemon.h

**Review by liorghub — CHANGES_REQUESTED** [review]

**Pull request closed without merging**